### ISSUE TITLE
Union type restrictions don't work when used within other objects

### ISSUE THREAD

**Issue by TiddoLangerak** [facebook/flow#3749]
Easiest to explain with code:

```jsx
type MyType = "A" | "B"
type WrappedType = {
  myType : MyType
}

const x : MyType = "A";
const y : WrappedType = {
  myType : x
}

if (x === 'C') { //Error: string literal `C`. This type is incompatible with string enum
  //...
}

if (y.myType === 'C') { //No error, even though this is just as invalid as the above
  // And from here `y` is treated as an `Object` type, meaning that any property access is valid:
  const z = y.x * 3; // No error, even though this is obviously invalid
}
```

It seems to me that the behavior of the second conditional is a bug. Or is there any reason why flow doesn't detect the errors here?

**Comment by mwalkerwells** [comment]
Believe this behavior is known & just a tradeoff until immutability features are added for objects (```Object.freeze``` / ```$Frozen```, etc) in Flow. Flow currently only recognizes that ```const x``` is an immutable reference.

Since projects are gradually typed, it's possible that ```y.myType``` is mutated / updated to ```'C'```.

@samwgoldman: https://github.com/facebook/flow/issues/739

**Comment by TiddoLangerak** [comment]
I don't think that this is related to immutability: changing all the `const` to `let` (or even `var`) still gives exactly the same behavior.

**Comment by madbence** [comment]
afaik accessing `y.myType` is expected behavior. but there is definitely inconsistency there.

**Comment by TiddoLangerak** [comment]
Could you elaborate why it's expected behaviour? @mwalkerwells mentioned that:

> Since projects are gradually typed, it's possible that y.myType is mutated / updated to 'C' elsewhere.

but if that is the reason, then there are still issues with this: flow does allow to reassign `y.myType` to a variable of type `MyType`, and from there it *will* complain about invalid comparison. E.g.:

```js
if (y.myType === 'C') { //No error
}
const extracted : MyType = y.myType;
if (extracted === 'C') { //error
}
```
If flow doesn't complain because `y.myType` could be mutated, then it should complain instead when trying to assign it to a `MyType` variable. 

I also discovered that it even goes further than that: even if I explicitly tell flow that `y.myType` is still a `MyType`, then it still won't error:

```js
if ((y.myType : MyType) === 'C') { // No error
}
```
